storage/replica.go

```diff
@@ -341,11 +341,27 @@ func (r *Replica) requestLeaderLease(timestamp roachpb.Timestamp) error {
 	// Note that the command itself isn't traced, but usually the caller
 	// waiting for the result has an active Trace.
 	errChan, pendingCmd := r.proposeRaftCommand(r.context(), ba)
-	if err := <-errChan; err != nil {
-		return err
+
+	// It does no good to block here for longer than the proposed lease's
+	// lifetime. If the command takes that long to commit we are probably
+	// partitioned away and we can conclude that another node has taken
+	// the lease.
+	deadline := time.After(duration)
```

Comment:
Using `context.WithDeadline` is only better than `time.After` if we give it a non-`Background` context (i.e. one that may already have a request deadline). And while we want to do that in many places, I don't think this is one of them: Thanks to the lock in `redirectOnOrAcquireLeaderLease`, one call to `requestLeaderLease` may serve many clients. If we bail out early because the first client expired or cancelled, other clients blocking on the lock could create redundant lease requests. We should wait here as long as the lease may be useful, regardless of the context.